api/client/gro_client.py

```diff
@@ -1,82 +1,167 @@
+class GroClient(Client):
+    """A Client with methods to find, and manipulate data series related
+    to a crop and/or region.
 
+    This class offers convenience methods for some common scenarios
+
+    - finding entities by name rather than ids
+    - exploration shortcuts filling in partial selections
+    - finding and saving data series for repeated use, including in a data frame
 
-def print_random_data_series(client, selected_entities):
-    """Example which prints out a CSV of a random data series that
-    satisfies the (optional) given selection.
-    """
-    # Pick a random data series for this selection
-    data_series_list = client.get_data_series(**selected_entities)
-    if not data_series_list:
-        raise Exception("No data series available for {}".format(
-            selected_entities))
-    data_series = data_series_list[int(len(data_series_list)*random())]
-    print("Using data series: {}".format(str(data_series)))
-    print("Outputing to file: {}".format(OUTPUT_FILENAME))
-    writer = unicodecsv.writer(open(OUTPUT_FILENAME, 'wb'))
-    for point in client.get_data_points(**data_series):
-        writer.writerow([point['start_date'], point['end_date'],
-                         point['value'] * point['input_unit_scale'],
-                         client.lookup_unit_abbreviation(point['input_unit_id'])])
-
-
-def search_for_entity(client, entity_type, keywords):
-    """Returns the first result of entity_type (which is items, metrics or
-    regions) that matches the given keywords.
-    """
-    results = client.search(entity_type, keywords)
-    for result in results:
-        print("Picking first result out of {} {}: {}".format(
-            len(results), entity_type, result['id']))
-        return result['id']
-    return None
-
-
-def pick_random_entities(client):
-    """Pick a random item that has some data associated with it, and a
-    random metric and region pair for that item with data
-    available.
     """
-    item_list = client.get_available('items')
-    num = 0
-    while not num:
-        item = item_list[int(len(item_list)*random())]
-        print("Randomly selected item: {}".format(item['name']))
-        selected_entities = {'itemId':  item['id']}
-        entity_list = client.list_available(selected_entities)
-        num = len(entity_list)
-    entities = entity_list[int(num*random())]
-    print("Using entities: {}".format(str(entities)))
-    selected_entities.update(entities)
-    return selected_entities
+
+    def __init__(self, api_host, access_token):
+        super(GroClient, self).__init__(api_host, access_token)
+        self._logger = lib.get_default_logger()
+        self._data_series_list = []  # all that have been added
+        self._data_series_queue = []  # added but not loaded in data frame
+        self._data_frame = None
+
+    ###
+    # Finding, indexing and loading multiple data series into a data frame
+    ###
+    def get_df(self):
```

Comment:
Wrapping these functions in a class is a breaking change.

The [cash_crops.py](https://github.com/gro-intelligence/api-client/blob/4b6116c51d41e1b20ebbeecad4d8b30d69d9ccb6/api/client/samples/crop_models/cash_crops.py#L25) example demonstrated importing functions from gro_client.

So this is failing now with

```
Traceback (most recent call last):
  File "cash_crops.py", line 25, in <module>
    from api.client.gro_client import get_df
ImportError: cannot import name get_df
```

Reply:
done

Reply:
Verified this fixes the cash_crops sample. Someone may have followed our example and imported other functions from there though. Need to do this for all of them?

`search_for_entity()`, `print_random_data_series()`, and `pick_random_entities()`

Reply:
done